packages/cozy-stack-client/src/DocumentCollection.js

```diff
@@ -168,10 +168,46 @@ class DocumentCollection {
     return this.stackClient.fetchJSON(
       'POST',
       path,
-      await this.toMangoOptions(selector, options)
+      this.toMangoOptions(selector, options)
     )
   }
 
+  /**
+   * Handle index creation if it is missing.
+   *
+   * When an index is missing, we first check if there is one with a different
+   * name but the same definition. If yes, we copy it to re-use the same index
+   * and remove the old one.
+   * If there is none, we create the new index.
+   *
+   *
+   * @param {object} selector - The mango selector
+   * @param {object} options - The find options
+   * @private
+   */
+  async handleMissingIndex(selector, options) {
+    const { indexedFields, partialFilter } = options
+    const existingIndex = await this.findExistingIndex(selector, options)
+    const indexName = getIndexNameFromFields(indexedFields)
+    if (!existingIndex) {
+      await this.createIndex(indexedFields, {
+        partialFilter,
+        indexName
+      })
+    } else {
+      try {
+        await this.copyIndex(existingIndex, indexName)
+        await this.destroyIndex(existingIndex)
+      } catch (error) {
+        // There might be a conflict error when 2 queries using the same index
+        // are run in parallel
+        if (!isDocumentUpdateConflict(error)) {
+          throw error
+        }
+      }
```

Comment:
I'm wondering if we should not put this work in a `migrateIndex(from, to)` method. This way, it'll be clear that we're doing this work for the purpose of a migration. 

It's clear ATM because we have the use case in mind, I don't know if in 6 months or 1 year, we'll remember that we have done that because we changed the way we named the index for the sake of the optimistic approach. 

I'm wondering if we should not explain "With the optimistic index creation approach, we are going to create new design documents." what / how is the optimistic approach in the commit description?

Reply:
Yes, it seems indeed clearer to add a `migrateIndex` method :+1:

Reply:
Method added with dc1c684 and commit description available in 72b64bf